spec/fastsync/vdd/fastsyncL2.tla

```diff
@@ -0,0 +1,528 @@
+    LET peerHeight == bPool.peerHeights[msg.peerId] IN     
+    IF /\ msg.peerId \in bPool.peerIds 
+       /\ MsgOnTime(msg, randomFlag)
+       /\ msg.height >= peerHeight  
+    THEN    \* a correct and timely response 
+        LET pHeights == [bPool.peerHeights EXCEPT ![msg.peerId] = msg.height] IN
+        [bPool EXCEPT !.peerHeights = pHeights]                   
+    ELSE RemovePeer(msg.peerId, bPool)   \* the peer has sent us duplicate status message or not on time or it was not solicited response
+        
+        
+HandleBlockResponse(msg, bPool, randomFlag) == 
+    LET h == msg.block.height IN
+    IF /\ msg.peerId \in bPool.peerIds
+       /\ bPool.blockStore[h] = NilBlock 
+       /\ bPool.pendingBlocks[h] = msg.peerId 
+       /\ MsgOnTime(msg, randomFlag)  
+    THEN LET bStore == [bPool.blockStore EXCEPT ![h] = msg.block] IN 
+         LET rBlocks == [bPool.receivedBlocks EXCEPT![h] = msg.peerId] IN   
+         [bPool EXCEPT 
+            !.blockStore = bStore,
+            !.receivedBlocks = rBlocks
+         ]  
+    ELSE RemovePeer(msg.peerId, bPool)
+                       
+
+MaxPeerHeight(bPool) == 
+    IF bPool.peerIds = {}
+    THEN 0 \* no peers, just return 0
+    ELSE CHOOSE max \in { bPool.peerHeights[p] : p \in bPool.peerIds }:
+            \A p \in bPool.peerIds: bPool.peerHeights[p] <= max \* max is the maximum
+    
+
+\* returns next height for which request should be sent.
+\* returns NilHeight in case there is no height for which request can be sent.
+FindNextRequestHeight(bPool) == 
+    LET S == {i \in Heights:  
+                /\ i >= bPool.height  
+                /\ i <= MaxPeerHeight(bPool)   
+                /\ bPool.blockStore[i] = NilBlock 
+                /\ bPool.pendingBlocks[i] = NilPeer} IN
+    IF S = {} 
+        THEN NilHeight
+    ELSE      
+        CHOOSE min \in S:  \A h \in S: h >= min
+
+\* Returns number of pending requests for a given peer.
+NumOfPendingRequests(bPool, peer) == 
+    LET peerPendingRequests == 
+        {h \in Heights: 
+            /\ h >= bPool.height 
+            /\ bPool.pendingBlocks[h] = peer
+            /\ bPool.blockStore[h] = NilBlock
+        }
+    IN 
+    Cardinality(peerPendingRequests)
+
+\* Returns peer that can serve block for given height. 
+\* Returns NilPeer in case there are no such peer.
+FindPeerToServe(bPool, h) == 
+    LET peersThatCanServe == { p \in bPool.peerIds: 
+                /\ bPool.peerHeights[p] >= h 
+                /\ NumOfPendingRequests(bPool, p) <= PEER_MAX_REQUESTS } IN
+    
+    LET pendingBlocks == 
+        {i \in Heights: 
+            /\ i >= bPool.height 
+            /\ bPool.pendingBlocks[i] /= NilPeer
+            /\ bPool.blockStore[i] = NilBlock
+        } IN
+    
+    IF \/ peersThatCanServe = {} 
+       \/ Cardinality(pendingBlocks) > TARGET_PENDING 
+    THEN NilPeer 
+    ELSE CHOOSE p \in peersThatCanServe: TRUE    \* TODO: Maybe introduce nondeterminism for peer selection
```

Comment:
This code implements a very specific strategy of assigning blocks to peers. Sure, we need it in the implementation. But are we going to reason about it at the spec level? In the end, this assignment strategy can be hidden under `CHOOSE` -- a function whose behavior is unknown.

Reply:
My intuition here is as follows. If you forget about `PEER_MAX_REQUESTS` and `TARGET_PENDING` and just pick an active peer (maybe `\E` is better than `CHOOSE` here), then you make the assignment strategy more abstract. You will have additional executions, but the spec will be simpler for the human and the model checker (it might be harder for tlc though).

Reply:
Minimal requirement is picking a peer that has reported higher height. I am not sure we want to be more abstract than that. This logic adds `PEER_MAX_REQUESTS` and `TARGET_PENDING` concepts in addition. I guess that `PEER_MAX_REQUESTS` and `TARGET_PENDING` can be seen as implementation detail so we can probably remove it from this high level spec, and then make this function simpler and more abstract (using CHOOSE).